src/looprestoration.rs

```diff
@@ -23,3 +30,327 @@ pub union LooprestorationParams {
     pub filter: Align16<[[int16_t; 8]; 2]>,
     pub sgr: LooprestorationParams_sgr,
 }
+
+#[derive(Copy, Clone)]
+#[repr(C)]
+pub struct Dav1dLoopRestorationDSPContext {
+    pub wiener: FnVersion,
+    pub sgr: FnVersion,
+}
+
+impl Dav1dLoopRestorationDSPContext {
+    pub unsafe fn call<BD: BitDepthFnLoopRestoration>(
+        &self,
+        filter: FilterFn,
+        mut p: *mut BD::Pixel,
+        stride: ptrdiff_t,
+        left: *const [BD::Pixel; 4],
+        mut lpf: *const BD::Pixel,
+        w: libc::c_int,
+        h: libc::c_int,
+        params: *const LooprestorationParams,
+        edges: LrEdgeFlags,
+        bd: BD,
+    ) {
+        use FilterFn::*;
+
+        let fn_version = match filter {
+            Wiener5 | Wiener7 => self.wiener,
+            Sgr5x5 | Sgr3x3 | SgrMix => self.sgr,
+        };
+
+        bd.call(
+            fn_version, filter, p, stride, left, lpf, w, h, params, edges,
+        )
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum FilterFn {
```

Comment:
Should this be `LoopRestorationFilterFn`?  Also, I was writing them with the `Fn` first to be more similar to how `fn` precedes the `fn` name in `fn foo`; what do you think about that?